### ISSUE TITLE
No BootOrder is set; firmware will attempt recovery

### ISSUE THREAD

**Issue by razic** [rhboot/efibootmgr#115]
Hello,

Thanks for all of the work around efibootmgr it's been really helpful to me. I boot Arch via EFISTUB.

I recently updated the BIOS on my mobo, and I'm experiencing some strange behavior:

1. When I `efibootmgr -uv` on a fresh boot, I will see `No BootOrder is set; firmware will attempt recovery`
2. If I create a boot entry with `efibootmgr --create`, I can run `efibootmgr -uv` without an error, see the entry in the command output, and also in the UEFI menu upon reboot.
3. If I boot up again, it goes back to number 1, and I get that error trying to list out the boot entries. Though, the entry does retain itself in the boot UEFI boot menu.

How could I debug this further?

Motherboard: ASUS PRIME X470-PRO
Motherboard BIOS Version: 5007
Kernel: 5.2.1-arch1-1-ARCHf82
efibootmgr version: 16

I tried downgrading my BIOS one version back, then two versions back. For some reason it wouldn't let me update two versions back, but it would let me update a single version back. Same problem though.

Thanks again :)

```
efibootmgr -v -v
Could not read variable 'BootNext': No such file or directory
error trace:
 efivarfs.c:258 efivarfs_get_variable(): open(/sys/firmware/efi/efivars/BootNext-8be4df61-93ca-11d2-aa0d-00e098032b8c): No such file or directory
 lib.c:139 efi_get_variable(): ops->get_variable failed: No such file or directory
Could not read variable 'BootCurrent': No such file or directory
error trace:
 efivarfs.c:258 efivarfs_get_variable(): open(/sys/firmware/efi/efivars/BootCurrent-8be4df61-93ca-11d2-aa0d-00e098032b8c): No such file or directory
 lib.c:139 efi_get_variable(): ops->get_variable failed: No such file or directory
Could not read variable 'Timeout': No such file or directory
error trace:
 efivarfs.c:258 efivarfs_get_variable(): open(/sys/firmware/efi/efivars/Timeout-8be4df61-93ca-11d2-aa0d-00e098032b8c): No such file or directory
 lib.c:139 efi_get_variable(): ops->get_variable failed: No such file or directory
Could not read variable 'BootOrder': No such file or directory
error trace:
 efivarfs.c:258 efivarfs_get_variable(): open(/sys/firmware/efi/efivars/BootOrder-8be4df61-93ca-11d2-aa0d-00e098032b8c): No such file or directory
 lib.c:139 efi_get_variable(): ops->get_variable failed: No such file or directory
 efibootmgr.c:364 read_order(): efi_get_variable failed: No such file or directory
No BootOrder is set; firmware will attempt recovery
Could not read variable 'MirrorCurrent': No such file or directory
error trace:
 efivarfs.c:258 efivarfs_get_variable(): open(/sys/firmware/efi/efivars/MirrorCurrent-7b9be2e0-e28a-4197-ad3e-32f062f9462c): No such file or directory
 lib.c:139 efi_get_variable(): ops->get_variable failed: No such file or directory
Could not read variable 'MirrorRequest': No such file or directory
error trace:
 efivarfs.c:258 efivarfs_get_variable(): open(/sys/firmware/efi/efivars/MirrorRequest-7b9be2e0-e28a-4197-ad3e-32f062f9462c): No such file or directory
 lib.c:139 efi_get_variable(): ops->get_variable failed: No such file or directory
```

```
~ ls -l /sys/firmware/efi/efivars
total 0
-rw-r--r-- 1 root root  192 Jul 18 16:55 ConIn-8be4df61-93ca-11d2-aa0d-00e098032b8c
-rw-r--r-- 1 root root   40 Jul 18 16:55 ConOut-8be4df61-93ca-11d2-aa0d-00e098032b8c
-rw-r--r-- 1 root root 7695 Jul 18 16:55 db-d719b2cb-3d3a-4596-a3bc-dad00e67656f
-rw-r--r-- 1 root root 3728 Jul 18 16:55 dbx-d719b2cb-3d3a-4596-a3bc-dad00e67656f
-rw-r--r-- 1 root root    5 Jul 18 16:55 DeploymentModeNv-97e8965f-c761-4f48-b6e4-9ffa9cb2a2d6
-rw-r--r-- 1 root root 2937 Jul 18 16:55 KEK-8be4df61-93ca-11d2-aa0d-00e098032b8c
-rw-r--r-- 1 root root 1369 Jul 18 16:55 PK-8be4df61-93ca-11d2-aa0d-00e098032b8c
-rw-r--r-- 1 root root   11 Jul 18 16:55 SecureBootSetup-7b59104a-c00d-4158-87ff-f04d6396a915
-rw-r--r-- 1 root root 5704 Jul 18 16:55 StdDefaults-4599d26f-1a11-49b8-b91f-858745cff824
```

**Comment by mologie** [comment]
This affects me too (most likely since I updated to BIOS 5007). I am using Ubuntu 19.04 with kernel 5.0.0-27. The recent ASUS BIOS update 5204 shows the same behavior.

I found that `mokutil` misbehaves and claims that secure boot is disabled, even though it is enabled. This causes dkms to fail horribly. I suspect that both efibootmgr and mokutil make use of EFI run-time variables, which the BIOS update somehow broke.

**Comment by mologie** [comment]
The following entry appears in my kernel logs:

`[   21.003939] efivars: duplicate variable: DeploymentModeNv-97e8965f-c761-4f48-b6e4-9ffa9cb2a2d6`

After investigating further in the kernel sources [0] I found that the kernel stops reading further variables after one duplicate was found. Thus I had the theory that somehow a dup. variable creeped in after the BIOS upgrade, and removing it might fix the issue.

Unfortunately, the EFI shell confirmed the duplicate but reported that the variable is write-protected, hence the error cannot be fixed by deleting it:

![image](https://user-images.githubusercontent.com/597682/64412028-0cc1e080-d07e-11e9-8a62-65f6ede60cc4.png)

Therefore I doubt that this is an issue in efibootmgr, but rather one in ASUS' firmware (and therefore the kernel, because it can work around the bug -- the EFI shell handles the duplicate properly after all).

[0] https://github.com/torvalds/linux/blob/master/drivers/firmware/efi/vars.c#L484

**Comment by last-partizan** [comment]
@mologie have you had any success fixing this issue?

**Comment by razic** [comment]
@mologie I also have that in my kernel logs:

```
➜ ~ dmesg | grep efivar
[    0.813585] Registered efivars operations
[    1.165475] efivars: duplicate variable: DeploymentModeNv-97e8965f-c761-4f48-b6e4-9ffa9cb2a2d6
```

**Comment by mologie** [comment]
I could not install a BIOS patch yet, because I found that my SPI flash programmer is not compatible with my board. Don't want to go all-in without a proper backup.

ASUS released a BIOS update in the meantime that mentions some Linux fixes. Unfortunately I found that it does not resolve the issue described in this thread. Chances are that we might also get this fixed by reaching right people at ASUS.

In the meantime I fixed stuff with a kernel patch, which works around the issue by retrying the read op if it encounters a duplicate. I'm not familiar with that kernel subsystem so this may have unintended side effects:

```patch
diff --git a/drivers/firmware/efi/vars.c b/drivers/firmware/efi/vars.c
index 436d1776bc7b..4714c79babb1 100644
--- a/drivers/firmware/efi/vars.c
+++ b/drivers/firmware/efi/vars.c
@@ -431,6 +431,8 @@ int efivar_init(int (*func)(efi_char16_t *, efi_guid_t, unsigned long, void *),
 	efi_char16_t *variable_name;
 	efi_status_t status;
 	efi_guid_t vendor_guid;
+	const int read_attempt_max = 16;
+	int read_attempt = 0;
 	int err = 0;
 
 	if (!__efivars)
@@ -449,17 +451,33 @@ int efivar_init(int (*func)(efi_char16_t *, efi_guid_t, unsigned long, void *),
 		goto free;
 	}
 
-	/*
-	 * Per EFI spec, the maximum storage allocated for both
-	 * the variable name and variable data is 1024 bytes.
-	 */
-
 	do {
-		variable_name_size = 1024;
+		/*
+		 * ASUS X470 boards with firmware 5007+ have a broken
+		 * get_next_variable implementation, where the same
+		 * variable is returned by multiple read calls. Work
+		 * around the issue by giving the function a few more
+		 * chances prior to erroring out.
+		 */
+
+		read_attempt = 0;
+
+		do {
+			/*
+			 * Per EFI spec, the maximum storage allocated for both
+			 * the variable name and variable data is 1024 bytes.
+			 */
+
+			variable_name_size = 1024;
+			read_attempt++;
+			status = ops->get_next_variable(&variable_name_size,
+							variable_name,
+							&vendor_guid);
+		} while (duplicates &&
+			 read_attempt < read_attempt_max &&
+			 status == EFI_SUCCESS &&
+			 variable_is_present(variable_name, &vendor_guid, head));
 
-		status = ops->get_next_variable(&variable_name_size,
-						variable_name,
-						&vendor_guid);
 		switch (status) {
 		case EFI_SUCCESS:
 			if (duplicates)
@@ -468,17 +486,18 @@ int efivar_init(int (*func)(efi_char16_t *, efi_guid_t, unsigned long, void *),
 			variable_name_size = var_name_strnsize(variable_name,
 							       variable_name_size);
 
+			if (read_attempt != 1) {
+				printk(KERN_WARNING "efivars: needed %d attempts "
+				       "to advance to next the EFI variable\n",
+				       read_attempt);
+			}
+
 			/*
-			 * Some firmware implementations return the
-			 * same variable name on multiple calls to
-			 * get_next_variable(). Terminate the loop
-			 * immediately as there is no guarantee that
-			 * we'll ever see a different variable name,
-			 * and may end up looping here forever.
+			 * Terminate the loop immediately as there is no
+			 * guarantee that we'll ever see a different variable
+			 * name, and may end up looping here forever.
 			 */
-			if (duplicates &&
-			    variable_is_present(variable_name, &vendor_guid,
-						head)) {
+			if (duplicates && read_attempt == read_attempt_max) {
 				dup_variable_bug(variable_name, &vendor_guid,
 						 variable_name_size);
 				status = EFI_NOT_FOUND;
```

**Comment by richscheller** [comment]
I am also seeing this problem on my Prime X470-Pro, with Ubuntu 19.04 (5.0.0-27 kernel) and the 5216 BIOS release.  Attempting to delete the duplicate variable in EFI shell still fails with "Write Protected".

I chatted online with ASUS support and opened a problem ticket to see if they (or AMI) will fix their BIOS in a future release, referencing this URL and pointing out that a BIOS developer needs to look into it.  I suggest that everyone seeing this issue do the same.  Multiple customer bug reports on the same issue have a lot more weight than just one bug report.

**Comment by twifty** [comment]
Exact same problem on the CROSSHAIR VII HERO (WIFI) with the 2801 BIOS.

**Comment by herobrauni** [comment]
Same issue on the ASUS OG-STRIX-X470-F-GAMING since upgrading to BIOS 5220.
I will try to downgrade and report back.

**Comment by mologie** [comment]
> Your patch works for me, but obv. requires manual compilation which is a little tedious.
> Did you have any success patching your BIOS?

@herobrauni I did not investigate this issue further yet and am running a custom Linux kernel build ever since. Thesis/university stuff eats up most of my time right now. Most likely I will get the chance to look into it December. However, I do recommend contacting ASUS support to make them more aware of the issue.

**Comment by herobrauni** [comment]
I am currently in contact with ASUS and they are looking into the problem.
Will update the comment if an update fixes the issue.

**Comment by hal3003** [comment]
> I am currently in contact with ASUS and they are looking into the problem.
> Will update the comment if an update fixes the issue.

Did you hear back from Asus?
I have an ASUS STRIX X470-F Gaming mainboard too and am experiencing the same issue.

**Comment by herobrauni** [comment]
> > I am currently in contact with ASUS and they are looking into the problem.
> > Will update the comment if an update fixes the issue.
> 
> Did you hear back from Asus?
> I have an ASUS STRIX X470-F Gaming mainboard too and am experiencing the same issue.

So far they were not able to give me any fix and tbh are quite slow to respond.

Feel free to open a ticket with them, the more the better.
If you want to reference mine, the ID is ASUS Service No=E1911121485

Should I get a fix I'll post it here.

**Comment by cscroggs** [comment]
Does anybody know if there is a resolution for this issue? I am having this exact issue with an ASUS X470 Prime Pro.

Does anybody know if BIOS 5406 solves the problem?

Besides the issue with efibootmgr what other problems might this bug cause? How concerned should I be?

**Comment by mologie** [comment]
> Does anybody know if there is a resolution for this issue? I am having this exact issue with an ASUS X470 Prime Pro.

I've reverse engineered relevant parts of the firmware, and the resolution appears to be to reset the NVRAM partition of the UEFI BIOS image. This requires a flash programmer. I have ordered one, and will report back with results as soon as it arrives here (~1 week).

Alternatively, you could contact ASUS support and RMA it. However, I will not go that route, because I expect to just receive it back as "working as intended". After all, Windows remains mostly unaffected by the issue.

> Does anybody know if BIOS 5406 solves the problem?

Unfortunately, it does not resolve the problem.

> Besides the issue with efibootmgr what other problems might this bug cause? How concerned should I be?

The primary issue is that Linux cannot be used to install EFI boot entries. Consequently, distro installations and upgrades will cause weird boot issues with will require expert knowledge, and possibly a Windows installation or UEFI shell, to fix.

**Comment by razic** [comment]
@mologie 

> The primary issue is that Linux cannot be used to install EFI boot entries.

Are you sure about that? efibootmgr still works for me. Even with this bug, as I mentioned in my original post:

> If I create a boot entry with efibootmgr --create, I can run efibootmgr -uv without an error, see the entry in the command output, and also in the UEFI menu upon reboot.

**Comment by mologie** [comment]
@razic

> Are you sure about that? efibootmgr still works for me. Even with this bug, as I mentioned in my original post:

Lucky you! ;) Efibootmgr works if the duplicate happens after Linux imported all critical variables required by it. In my case, it happens before importing the Boot* entries, efibootmgr fails because the entries are invisible to it. Of course, the EFI shell is unaffected, because it handles duplicated entries just fine where as Linux explicitly errors out.

**Comment by razic** [comment]
That makes sense :P Guess I did get lucky. Sucks that this bug is worse than I had thought for others though.

**Comment by cscroggs** [comment]
I'm trying to determine my options since it seems unlikely that a fix is going to come from ASUS. Do you think a change to work around duplicate variables would ever make into the mainline kernel?

I've learned more about EFI than I ever intended but I'm still feeling my way around. Does using something like rEFInd help the situation any?

Should I just cut my losses and replace the board? Not ideal but it seems this issue could be around for the life of the board.

**Comment by mologie** [comment]
@cscroggs 

> I'm trying to determine my options since it seems unlikely that a fix is going to come from ASUS. Do you think a change to work around duplicate variables would ever make into the mainline kernel?

Unlikely, because it's a really dirty hack fix for what could be considered defect hardware. One would also have to get the maintainers on board. I have, for obvious reasons, not supported the patch to the Linux kernel-dev list.

> I've learned more about EFI than I ever intended but I'm still feeling my way around. Does using something like rEFInd help the situation any?

REFInd itself probably not, but I imagine it's possible to program an EFI runtime service, which hides the duplicate variables from the OS. Bootloaders for macOS, such as OpenCore, use comparable methods. Then again, this is a custom hackfix for broken firm-/hardware and I'd rather just fix the hardware with a flash programmer.

> Should I just cut my losses and replace the board? Not ideal but it seems this issue could be around for the life of the board.

That would be the easiest solution indeed. I'll check back in here next week once I got my flash programmer and tested fixing up the UEFI flash. Basically, I would thereby reset the board to factory defaults -- an option that ASUS for some reason does not provide at all via software. So this is up to you: Either repurpose it (with Windows / custom kernel), RMA it, or take your chances with a flash programmer.

**Comment by mologie** [comment]
Good news, everyone!

Flashing the board to a factory image worked fabulously to fix the duplicated EFI variable errors. I spent a few hours on this to figure out how everything worked, but I guess that with a guide, you'll get away with a multimeter, $10 of hardware from China, and half an hour of your time.

Disclaimer: This applies only to ASUS X470-F boards. Double and triple check everything you do here. AM4 wants 1.8V for the SPI flash, and connecting a 3.3V/5V flashing device will fry your board. Keep backups of your old SPI flash contents just in case.

The short version is:

1. Get a CH341A (black board) or compatible SPI flash programmer.
2. Get a 1.8V SPI flash adapter for the above -- these are often sold with the programmer as bundle.
3. Disconnect power from the board and wire everything up. The 9-pin header next to the SPI flash chip on the board can be used for connecting to it. This is where your multimeter comes in handy, and I will elaborate it in my blog post later, but to be brief the mapping is: NC 3 8 4 1 6 2 5 7. Numbers are chip pins counter-clockwise like shown on page 7 of the datasheet [1].
4. Use flashrom (latest master -- the stable release refuses to work with the chip) to dump your BIOS contents.
5. Download the latest BIOS CAP image from ASUS.
6. Strip the AMI capsule header with dd (remove the first 4096 bytes) to the the master ROM image
7. Customize the ROM image for your board by using FD44Editor [2]: Read in your BIOS dump, save to the master ROM you extracted from the CAP file.
8. Write the personalized ROM file back to the SPI chip with flashrom.
9. Tada! You are now in possession of either a brick (please don't sue me, though my lawyer loves those cases), or a factory-reset board with latest firmware (congratulations)!

I will write a blog post with an elaborate guide Soon(tm). Until then, good luck!

![20200216_184328787_iOS](https://user-images.githubusercontent.com/597682/74611475-71d53400-50fc-11ea-97f6-bac706f6548b.jpeg)

[1] https://www.macronix.com/Lists/Datasheet/Attachments/7634/MX25U25645G,%201.8V,%20256Mb,%20v1.3.pdf

[2] https://github.com/LongSoft/FD44Editor/releases

**Comment by cscroggs** [comment]
Thanks for the information.

I've not done anything like this before. Will this be a permament fix as long as I don't install another BIOS update from ASUS? I've always been a little unclear about where this duplicate came from. Was it in the BIOS from the factory or did I do something which generated? If the latter, how does this procedure prevent that?

There are couple of steps, especially steps six, seven and eight. I'm looking forward to your more detailed blog post.

**Comment by mologie** [comment]
@cscroggs 

> Will this be a permament fix as long as I don't install another BIOS update from ASUS? I've always been a little unclear about where this duplicate came from. Was it in the BIOS from the factory or did I do something which generated? If the latter, how does this procedure prevent that?

I do not know the cause of the data corruption. It is definitely not present in the latest factory images, because these contain an empty UEFI variable partition. The partition is (correctly) initialized on first run of the firmware.

It's reasonable to assume that a firmware upgrade itself caused data corruption, because I cannot find any reports on the internet of people who have never upgraded firmware and yet experience this issue. The fix is likely to be permanent, unless the upgrade/migration routine that breaks stuff runs on every upgrade.

FWIW, I have used ASUS' EasyFlash tool built into the UEFI firmware to "upgrade" to the same firmware version, and the bug did not happen again. Neither has it happened randomly on my machine since I flashed a new factory image. We will have to wait for the next update from ASUS to see whether the fix is indeed permanent.

tl;dr: It's most likely a permanent fix

**Comment by sclem** [comment]
@mologie thank you for all the info. I have a Prime x470-Pro mobo and was able to fix the efibootmgr issues with your kernel patch on arch. I'm not confident enough to flash the board yet and the patch works well enough I am willing to recompile the kernel for updates. Do you know if it would be possible to load this patch as a dkms module? If so we could make a asus-x470-efi-dkms package to help other users.

**Comment by mologie** [comment]
@sclem most likely this cannot be loaded as dkms module. We are modifying an existing module that is already compiled into the kernel, and to the best of my knowledge dynamic kernel modules cannot replace statically linked ones.

**Comment by kdarkhan** [comment]
I believe this issue was resolved in the latest bios. I just installed version 5837 from 2021/02/09 and the issue is gone on my Prime x470-Pro. Around 6 months ago I created an issue through their customer support. I was trying to persuade them that this was a bug in their firmware but they did not bulge. They were convincing me to reinstall my windows 🤯 
I gave up. I guess Asus realized that it was their problem after all.

**Comment by sclem** [comment]
I can confirm bios version 5837 fixes this issue for me as well.

**Comment by spacz80** [comment]
I have just update to BIOS version 5837 but the problem is still there. For now, I continue to use the @mologie kernel patch which is fixing the issue for me.

**Comment by NevilleShorttop** [comment]
@mologie 
> * Get a CH341A (black board) or compatible SPI flash programmer.
> 
> * Get a 1.8V SPI flash adapter for the above -- these are often sold with the programmer as bundle.
> 
> * Disconnect power from the board and wire everything up. The 9-pin header next to the SPI flash chip on the board can be used for connecting to it. This is where your multimeter comes in handy, and I will elaborate it in my blog post later, but to be brief the mapping is: NC 3 8 4 1 6 2 5 7. Numbers are chip pins counter-clockwise like shown on page 7 of the datasheet [1].
> 
> * Use flashrom (latest master -- the stable release refuses to work with the chip) to dump your BIOS contents.
> 
> * Download the latest BIOS CAP image from ASUS.
> 
> * Strip the AMI capsule header with dd (remove the first 4096 bytes) to obtain a master/factory ROM image.
> 
> * Customize the ROM image for your board by using FD44Editor [2]: Read in your BIOS dump, save to the ROM you extracted from the CAP file.
> 
> * Write the personalized ROM file back to the SPI chip with flashrom.
> 
> * Tada! You are now in possession of either a brick (please don't sue me, though my lawyer loves those cases), or a factory-reset board with latest firmware (congratulations)!

This worked perfectly. Thanks!

**Comment by tietzc** [comment]
Finally fixed for me on X470-F Gaming since BIOS version 6024.

**Comment by frozencemetery** [comment]
It is awesome to see the stuff you did around this, @mologie et al..  That said, it sounds like a fix has been provided by Asus (and also, there's nothing efibootmgr can do about it anyway) so I'm closing this out.

**Comment by mologie** [comment]
Thanks for providing the forum here! Glad to hear that apparently there is an update to fix this now. I hope that this really is it since the changelog is vague as always, but you can only hope. I am not spending time to take this one apart... probably ;-).

**Comment by SaphireLattice** [comment]
Just updated my firmware on ASUS X470 Prime to 6042 after having missing BootCurrent/etc, and... Nothing is fixed :) Fuck.

And the variable is very distinctly write protected still, etc

Guess I'll be waiting until I can get a flasher and do the thing described above... Sigh. Or patch the kernel to be less dumb about all of this...
